lib/addressfinder/v1/email/batch_verification.rb

```diff
@@ -0,0 +1,69 @@
+module AddressFinder
+  module V1
+    module Email
+      class BatchVerification
+        attr_reader :emails, :results
+
+        # Verifies an array of email addresses using concurrency to reduce the execution time.
+        # The results of the verification are stored in the `results` attribute, in the same order
+        # in which they were supplied.
+        #
+        # @param [Array<String>] emails
+        # @param [AddressFinder::HTTP] http HTTP connection helper
+        # @param [Integer] concurrency How many threads to use for verification
+        # @param [Hash] args Any additional arguments that will be passed onto the EV API
+        def initialize(emails:, http:, concurrency: 5, **args)
+          @emails = emails
+          @concurrency = concurrency
+          @http = http
+          @args = args
+        end
+
+        def perform
+          confirm_concurrency_level
+          verify_each_email_concurrently
+
+          self
+        end
+
+        private
+
+        attr_reader :args, :concurrency, :http
+
+        MAX_CONCURRENCY_LEVEL = 10
+
+        def confirm_concurrency_level
+          return unless @concurrency > MAX_CONCURRENCY_LEVEL
+
+          warn "WARNING: Concurrency level of #{@concurrency} is higher than the maximum of #{MAX_CONCURRENCY_LEVEL}. Using #{MAX_CONCURRENCY_LEVEL}."
+          @concurrency = MAX_CONCURRENCY_LEVEL
+        end
+
+        def verify_each_email_concurrently
+          @results = Concurrent::Array.new(emails.length)
+
+          pool = Concurrent::FixedThreadPool.new(concurrency)
+
+          @emails.each_with_index do |email, index_of_email|
+            # Start a new thread for each task
+            pool.post do
+              verify_email(email, index_of_email)
+            end
+          end
+
+          ## Shutdown the pool and wait for all tasks to complete
+          pool.shutdown
+          pool.wait_for_termination
+        end
```

Comment:
All the way until here, this is basically verification agnostic. I wonder if it is worth supporting concurrent calls for both Phone and physical addresses from the get go? possibly similar to what the bulk API does, and passing the verification class to be used for verification?
this could be extracted later on too

Reply:
I've taken a look at how the Portal calls this gem, and it looks like it would be easier to add the AV and PV batch verification in a similar manner, such that the Portal code can continue to also use the same algorithm for all three data types. 

I'll be creating a card to add the additional data types for concurrency treatment next week.